### PR TITLE
Merge attach and detach common func

### PR DESCRIPTION
**Pull request by cwdsuzhou**
**What type of PR is this?**

/kind cleanup
/kind feature

**What this PR does / why we need it**:

Since `waitForVolumeAttachmentStatus` and `waitForVolumeDetachmentStatus` have similar logic, we'd better merge them.

**Which issue(s) this PR fixes**:

Fixes #

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note
NONE
```

**Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.**:

```docs

```

### CODE REVIEW

**Comment by cwdsuzhou** [comment]
@jsafrane  please take a look at this PR, thanks.

**Comment by jsafrane** [comment]
I'd prefer if #79144 is merged first, you can then rebase (and look what's left there).

**Comment by cwdsuzhou** [comment]
> I'd prefer if #79144 is merged first, you can then rebase (and look what's left there).

Thanks.

/hold

**Comment by cwdsuzhou** [comment]
@jsafrane https://github.com/kubernetes/kubernetes/pull/79144 has merged, PLAT this PR now

File: pkg/volume/csi/csi_attacher.go [hunk 1]
```diff
@@ -51,6 +51,10 @@ type csiAttacher struct {
 	csiClient csiClient
 }
 
+type waitForStatus func(attach *storage.VolumeAttachment, volumeHandle string) (bool, error)
+
+type waitForStatusInternal func(volumeHandle, attachID string, timer *time.Timer, timeout time.Duration) (string, error)
```
Comment (jsafrane) [inline pkg/volume/csi/csi_attacher.go]: This type it not used (well, it is used, but parameter of this type is not used).

File: pkg/volume/csi/csi_attacher.go [hunk 2]
```diff
@@ -458,27 +416,46 @@ func (c *csiAttacher) waitForVolumeDetachment(volumeHandle, attachID string) err
...
-		klog.Error(log("detachment for VolumeAttachment [%v] for volume [%s] failed: %v", attachID, volumeHandle, detachErr.Message))
-		return errors.New(detachErr.Message)
+		klog.Error(log("detachment for VolumeAttachment for volume [%s] failed: %v", volumeHandle, detachErr.Message))
+		return false, errors.New(detachErr.Message)
+	}
+	return false, nil
+}
+
+func (c *csiAttacher) waitForVolumeAttachDetachmentStatus(attach *storage.VolumeAttachment, volumeHandle, attachID string,
+	timer *time.Timer, timeout time.Duration, waitForStatus waitForStatus, waitForReadyInternal waitForStatusInternal) (string, error) {
```
Comment (jsafrane) [inline pkg/volume/csi/csi_attacher.go]: waitForReadyInternal is not used

Comment (jsafrane) [inline pkg/volume/csi/csi_attacher.go]: `waitForStatus` is quite confusing name, because it does not wait for anything.

Comment (jsafrane) [inline pkg/volume/csi/csi_attacher.go]: Only `attachID` is returned as the `string` return value. Can this function just return error and caller can use `attachID` when `waitForVolumeAttachDetachmentStatus` returns success?

File: pkg/volume/csi/csi_attacher.go [hunk 3]
```diff
@@ -458,27 +416,46 @@ func (c *csiAttacher) waitForVolumeDetachment(volumeHandle, attachID string) err
...
 	if detachErr != nil {
-		klog.Error(log("detachment for VolumeAttachment [%v] for volume [%s] failed: %v", attachID, volumeHandle, detachErr.Message))
-		return errors.New(detachErr.Message)
+		klog.Error(log("detachment for VolumeAttachment for volume [%s] failed: %v", volumeHandle, detachErr.Message))
+		return false, errors.New(detachErr.Message)
+	}
+	return false, nil
+}
+
+func (c *csiAttacher) waitForVolumeAttachDetachmentStatus(attach *storage.VolumeAttachment, volumeHandle, attachID string,
```
Comment (jsafrane) [inline pkg/volume/csi/csi_attacher.go]: Either `waitForVolumeAttachDetachStatus` or `waitForVolumeAttachmentDetachmentStatus`, having it half way looks odd.

File: pkg/volume/csi/csi_attacher.go [hunk 4]
```diff
@@ -488,30 +465,31 @@ func (c *csiAttacher) waitForVolumeDetachmentInternal(volumeHandle, attachID str
...
+					return "", err
+				}
+				if successful {
+					return attachID, nil
 				}
 			case watch.Deleted:
 				//object deleted
 				klog.V(4).Info(log("VolumeAttachment object [%v] for volume [%v] has been deleted", attachID, volumeHandle))
-				return nil
+				return attachID, nil
```
Comment (jsafrane) [inline pkg/volume/csi/csi_attacher.go]: This is correct when we wait for detach (i.e. VolumeAttachment being deleted). It should be error when we're waiting for attach.

**Comment by cwdsuzhou** [comment]
@jsafrane thanks for your careful review. I have change the PR according to your comments.

PTAL

File: pkg/volume/csi/csi_attacher.go [hunk 5]
```diff
@@ -445,10 +386,11 @@ func (c *csiAttacher) Detach(volumeName string, nodeName types.NodeName) error {
 	}
 
 	klog.V(4).Info(log("detacher deleted ok VolumeAttachment.ID=%s", attachID))
-	return c.waitForVolumeDetachment(volID, attachID)
+	_, err := c.waitForVolumeDetachment(volID, attachID)
+	return err
 }
 
-func (c *csiAttacher) waitForVolumeDetachment(volumeHandle, attachID string) error {
+func (c *csiAttacher) waitForVolumeDetachment(volumeHandle, attachID string) (string, error) {
```
Comment (jsafrane) [inline pkg/volume/csi/csi_attacher.go]: Returning `string` looks useless, caller ignores it. It applies also to `waitForVolumeDetachmentInternal` below - nobody needs attachID return value during detach.

File: pkg/volume/csi/csi_attacher.go (hunk 4 above)
Comment (jsafrane) [reply pkg/volume/csi/csi_attacher.go]: There is still a bug, IMO. Current code in `case watch.Deleted:` works for detachment - a volume has been detached when VolumeAttachment is deleted.

It does not work for attachment - something went wrong when `VolumeAttachment` got deleted and the volume is definitely not attached. It must not return success in this case.

* You can either add extra parameter to `waitForVolumeAttachmentInternal` to differentiate between attach/detach and return proper return value on delete event .
* Or you can run `verifyStatus(nil, volumeHandle)`, where `nil` means that the attachment has been deleted (don't forget to document this).
* Or you can find a better solution, of course.

File: pkg/volume/csi/csi_attacher.go (hunk 5 above)
Comment (cwdsuzhou) [reply pkg/volume/csi/csi_attacher.go]: Fixed

File: pkg/volume/csi/csi_attacher.go (hunk 4 above)
Comment (cwdsuzhou) [reply pkg/volume/csi/csi_attacher.go]: I chose the 2rd method and added some comments.

PTAL, thanks

**Comment by cwdsuzhou** [comment]
PR changed according to the comments.

I would like open another PR to add some more unit tests.

File: pkg/volume/csi/csi_attacher.go [hunk 6]
```diff
@@ -494,24 +448,31 @@ func (c *csiAttacher) waitForVolumeDetachmentInternal(volumeHandle, attachID str
...
+				if successful {
+					return nil
 				}
 			case watch.Deleted:
-				//object deleted
+				// set attach nil to get different results
+				// for detachment, a deleted event means successful detachment, should return success
+				// for attachment, should return fail
+				if successful, err := verifyStatus(nil, volumeHandle); !successful {
+					klog.Error(log("VolumeAttachment [%s] has been deleted, will not continue to wait for attachment", volumeHandle))
```
Comment (jsafrane) [inline pkg/volume/csi/csi_attacher.go]: IMO, logging the error in `verifyAttachmentStatus` would be better, as other attach errors are logged there.

File: pkg/volume/csi/csi_attacher.go [hunk 7]
```diff
@@ -467,18 +410,29 @@ func (c *csiAttacher) waitForVolumeDetachmentInternal(volumeHandle, attachID str
 			klog.V(4).Info(log("VolumeAttachment object [%v] for volume [%v] not found, object deleted", attachID, volumeHandle))
 			return nil
 		}
-		return errors.New(log("detacher.WaitForDetach failed for volume [%s] (will continue to try): %v", volumeHandle, err))
+		klog.Error(log("detacher.WaitForDetach failed for volume [%s] (will continue to try): %v", volumeHandle, err))
```
Comment (jsafrane) [inline pkg/volume/csi/csi_attacher.go]: Is this necessary to log? The error should be already logged.

Comment (cwdsuzhou) [reply pkg/volume/csi/csi_attacher.go]: done

File: pkg/volume/csi/csi_attacher.go (hunk 6 above)
Comment (cwdsuzhou) [reply pkg/volume/csi/csi_attacher.go]: yeap , done

Thanks

**Comment by jsafrane** [comment]
/approve
/lgtm

Thanks for this PR!

**Comment by cwdsuzhou** [comment]
> [APPROVALNOTIFIER] This PR is **APPROVED**
> 
> This pull-request has been approved by: _[cwdsuzhou](https://github.com/kubernetes/kubernetes/pull/80420#)_, _[jsafrane](https://github.com/kubernetes/kubernetes/pull/80420#issuecomment-536506197)_
> 
> The full list of commands accepted by this bot can be found [here](https://go.k8s.io/bot-commands?repo=kubernetes%2Fkubernetes).
> 
> The pull request process is described [here](https://git.k8s.io/community/contributors/guide/owners.md#the-code-review-process)



> /approve
> /lgtm
> 
> Thanks for this PR!

Thanks for your careful review! :)

**Comment by cwdsuzhou** [comment]
> [APPROVALNOTIFIER] This PR is **APPROVED**
> 
> This pull-request has been approved by: _[cwdsuzhou](https://github.com/kubernetes/kubernetes/pull/80420#)_, _[jsafrane](https://github.com/kubernetes/kubernetes/pull/80420#issuecomment-536506197)_
> 
> The full list of commands accepted by this bot can be found [here](https://go.k8s.io/bot-commands?repo=kubernetes%2Fkubernetes).
> 
> The pull request process is described [here](https://git.k8s.io/community/contributors/guide/owners.md#the-code-review-process)



> /approve
> /lgtm
> 
> Thanks for this PR!

Thanks for your careful review! :)

**Pull request merged**